src/lib/db.js

```diff
@@ -26,22 +28,40 @@ const TABLE_NAMES = [
   'L_LU_TRANSFER_STATUS'
 ]
 
-function withTransaction (callback) {
-  return knex.transaction(co.wrap(callback))
+// this function takes a generator as its callback
+// and returns a promise
+function withTransaction (callbackGenerator) {
+  return knex.transaction(co.wrap(callbackGenerator))
 }
 
-function withSerializableTransaction (callback) {
+// this function takes a generator as its callback
+// and returns a promise
+function withSerializableTransaction (callbackGenerator, retries = 5) {
   const dbType = knex.client.config.client
-  return withTransaction(function * (transaction) {
-    // Set isolation level to avoid reading "prepared" transaction that is currently being
-    // executed by another request. This ensures the transfer can be fulfilled only once.
-    assert(_.includes(['sqlite3', 'pg', 'mysql'], dbType),
-      'A valid client must be specified on the db object')
-    if (dbType === 'pg') {
-      yield transaction.raw('SET TRANSACTION ISOLATION LEVEL SERIALIZABLE')
-    }
+  return promiseRetry(function (retry, attemptNo) {
+    return withTransaction(function * (transaction) {
+      // Set isolation level to avoid reading "prepared" transaction that is currently being
+      // executed by another request. This ensures the transfer can be fulfilled only once.
+      assert(_.includes(['sqlite3', 'pg', 'mysql'], dbType),
+        'A valid client must be specified on the db object')
+      if (dbType === 'pg') {
+        yield transaction.raw('SET TRANSACTION ISOLATION LEVEL SERIALIZABLE')
+      }
 
-    return yield callback(transaction)
+      yield * callbackGenerator(transaction)
+    }).catch(err => {
+      // 40001 is a postgres error code meaning the database could not complete the transaction
+      // because this would interfere with other concurrent transactions
+      if (err.code === '40001') {
+        log.debug('retrying database query', `${attemptNo}/${retries}`, err)
+        retry(err)
+      }
+      throw err
+    })
+  }, {
+    minTimeout: 0,
```

Comment:
By setting `minTimeout: 0` you retry immediately after the transaction fails. Is this intended? Would it make sense waiting for a little while before retrying?

For fulfillments I can see why you want to do it as fast as possible, but what about other transactions?